corehq/apps/sso/backends.py

```diff
@@ -76,8 +77,26 @@ def authenticate(self, request, username, idp_slug, is_handshake_successful):
             is_new_user = False
             web_user = WebUser.get_by_username(username)
         except User.DoesNotExist:
-            user, web_user = self._create_new_user(request, username, async_signup)
-            is_new_user = True
+            try:
+                # the atomic block is necessary to avoid any race conditions
+                # that came up during testing of this feature.
+                with transaction.atomic():
+                    user, web_user = self._create_new_user(request, username, async_signup)
+                    is_new_user = True
```

Comment:
If `self._create_new_user()` did need an atomic block, it should be inside the method, not enclosing it. The atomic block is meant to ensure multiple changes all happen together, and here, the only thing that is happening is that a new user is created. It's possible that the operations in `_create_new_user()` need to all happen together, but then that means I'd never want to call `_create_new_user()` _without_ an atomic block, due to the possibility of another race condition. We eliminate that possibility for all callers by moving the atomic block to where the multiple operations happen.